sympy/solvers/inequalities.py

```diff
@@ -549,9 +549,10 @@ def valid(x):
                 if v.is_extended_real is False:
                     return S.false
                 else:
-                    v = v.n(2)
                     if v.is_comparable:
                         return expr.func(v, 0)
+                     if v.equals(0):
+                        return expr.func(0, 0) 
```

Comment:
We should not use `.equals` here.

Reply:
This is a higher level routine and use of simplification is not prohibited when needed to get a solution. Why do you think it should not be used?

Reply:
There are bigger problems here like the fact that there is no definition of what cases solve_univariate_inequality should be expected to handle or what outputs it should return. Attempting to work around those with simplify will only make those worse. In fact attempting to improve anything about these inequalities solvers without first addressing the problem of defining what they are even supposed to do is a mistake.

Reply:
The name implies that it should solve the inequality for a symbol if possible. This is an easy equality to solve but proving that 9 is a solution requires showing that `log(9)/log(3) == 2`. That can't happen with out `expand_log`. Constraints of the `A=B`  problem being acknowledged, we cannot refrain from returning a solution for something that is easily shown to be a solution just because it might fail in general (and return None). Nor should every routine that finds itself in the `A - B = 1e-126` sort of dead-end internally have to re-code the various simplifications that might be needed to show that the `A - B` actually *is* 0. There is nothing that can be simplified about the original expression `log(x, 3) <= 2`. Autoexpansion to `log(x)/log(3) <= 2` makes the substitution of 9 for `x` fail evaluation because the context (that the base is 3) is now lost. And even if we solve this problem at a higher level, the un-ending possibilities remain and `equals` is the last defense against repeating the heuristics throughout the solve mechanics. 

`equals` is not the right thing (as you have instructed) for things like `is_positive` which cannot risk getting trapped in a slow evaluation. It is the right thing for a `solve` routine (and the docstring indicates that it should return solutions for nonlinear inequalities.

Reply:
> The name implies that it should solve the inequality for a symbol if possible

What does "solve" mean and what does "if possible" mean. The expected behaviour of this function is entirely unclear to me.

> we cannot refrain from returning a solution for something that is easily shown to be a solution just because it might fail in general (and return None

Why not?

Why can't the inequality be returned so that the user can call simplify if they want?

The documented specification of this function gives no answer as to what should be expected. The fact that things might fail in general is important: it would be much better if this function had a defined restricted scope and could be expected to "solve" all problems within that scope.

Reply:
> What does "solve" mean and what does "if possible" mean. 

This is the same question that is raised with `simplify`. `fu` comes the closest to saying what that means because they offer a complex series of simplification schemes to select an answer. In `simplify` we have a series of steps (which might be modified) that can be used to simplify expressions and possibly show that an expression is equal to zero. So for `sympy`, "solve if possible" means, "with the heuristics we employ we find a solution and return the evaluated solution if it can be coaxed to evaluation with the simplification methods. If not, we will either raise an error or return an unevaluated solution."

> Why can't the inequality be returned so that the user can call simplify if they want?

I hadn't considered returning `log(9)/log(3) <= 2` as a solution since I am usually thinking about selecting a solution to an equality rather than an inequality. In this case, that might be reasonable. If there is more than one solution then an `Or` would be returned as the unevaluated solution. I am not sure the same could be done with simplify.